### PR TITLE
Enable shadow warnings on GCC and tweak CMake 

### PR DESCRIPTION
**Pull request by AaronDMarasco**
When I was trying to squash #4806 I accidentally closed the previous PR.

Added new CMake macros to reduce explicit version/compiler checks for features

Found one unused variable (because the lambda shadowed, original never used)

Fixed one shadowed variable (but in debug mode will error if new code introduces them)

## Testing strategy
Locally tested with GCC 10.1.1 and clang 10.0.0 (Fedora 32). Taking advantage of CI system for other compilers.

Local cmake is 3.17.3 but didn't use any advanced features.

## Type of change
- ✅ Bug fix (non-breaking change that fixes an issue)

### CODE REVIEW

**Comment by AaronDMarasco** [comment]
> https://ci.keepassxc.org/viewLog.html?buildId=58396&buildTypeId=KeepassXC_UbuntuLinux
> TestCli::testClip() QTestLib: This test case check ("((((clipboard->text())) == (QString("User Name"))))") failed because the requested timeout (5000 ms) was too short, 5000 ms would have been sufficient this time.

I have no idea what that means and am 99.44% sure it has nothing to do with my code.

**Comment by droidmonkey** [comment]
It is just a CI hiccup

File: src/fdosecrets/objects/Collection.cpp [hunk 1]
```diff
@@ -490,7 +490,6 @@ namespace FdoSecrets
         // When the group object is normally deleted due to ~Database, the databaseReplaced
         // signal should be first emitted, and we will clean up connection in reloadDatabase,
         // so this handler won't be triggered.
-        QPointer<Database> db = m_backend->database().data();
```
Comment (AaronDMarasco) [inline src/fdosecrets/objects/Collection.cpp:493]: This was unused; the lambda was shadowing it with its own `db` which turned out to be the only one needed.

Comment (phoerious) [reply src/fdosecrets/objects/Collection.cpp:493]: Right, never mind. I misread the line.

**Pull request merged**